lib/plugins/cpuUsageThrottle.js

```diff
@@ -0,0 +1,139 @@
+ *    rejecting/accepting connections. It's best to keep this value as low as
+ *    possible without creating a significant impact on perfomance in order to
+ *    keep your server as responsive to change as possible.
+ * @param {Number} opts.halfLife When we sample the CPU usage on an interval,
+ *    we create a series of data points. We take these points and calculate a
+ *    moving average. The halfLife indicates how quickly a point "decays" to
+ *    half it's value in the moving average. The lower the halfLife, the more
+ *    impact newer data points have on the average. If you want to be extremely
+ *    responsive to spikes in CPU usage, set this to a lower value. If you want
+ *    your process to put more emphasis on recent historical CPU usage when
+ *    determininng whether it should shed load, set this to a higher value. The
+ *    unit is in ms. The default is to set this to the same value as interval.
+ * @param {Error} [opts.err] A restify error used as a response when the
+ *    cpu usage limit is exceeded
+ * @returns {Function} middleware to be registered on server.pre
+ */
+function cpuUsageThrottle (opts) {
+
+    // Scrub input and populate our configuration
+    assert.object(opts, 'opts');
+    assert.number(opts.limit, 'opts.limit');
+    assert.optionalNumber(opts.max, 'opts.max');
+    assert.number(opts.interval, 'opts.interval');
+    assert.optionalNumber(opts.halfLife, 'opts.halfLife');
+
+    if (opts.err !== undefined && opts.err !== null) {
+        assert.ok(opts.err instanceof Error, 'opts.err must be an error');
+        assert.optionalNumber(opts.err.statusCode, 'opts.err.statusCode');
+    }
+
+    var self = {};
+    self._err = opts.err || defaultResponse;
+    self._limit = opts.limit;
+    self._max = opts.max || 1;
+    self._interval = opts.interval;
+    self._halfLife = (typeof opts.halfLife === 'number') ?
+      opts.halfLife : opts.interval;
+    self._timeout = null;
+    assert.ok(self._max > self._limit, 'limit must be less than max');
+
+    // We start at 0 to compensate for the initial spike in CPU usage a server
+    // experiences when starting up.
+    self._ewma = new EWMA(self._halfLife, 0);
+
+    // self._reject represents the % of traffic that we should reject at the
+    // current point in time based on how much over our limit we are. This is
+    // updated on an interval by updateReject().
+    self._reject = 0;
+
+    // updateReject should be called on an interval, it checks the average CPU
+    // usage between two invocations of updateReject.
+    function updateReject() {
+        pidusage.stat(process.pid, function (e, stat) {
+            // If we fail to stat for whatever reason, err on the side of
+            // accepting traffic, while failure probably indicates something
+            // horribly wrong with the server, identifying this error case and
+            // appropriately responding is way outside the scope of this plugin
```

Comment:
Is it worth emitting an error in this case? Might be a way to hook up metrics for when the plugin is misbehaving for whatever reason.

Reply:
The plugin reads the metrics from either stdout or a file descriptor, the error we would get here would be the same as a read from the filesystem failing. Something horribly wrong is happening, and more than likely is showing up somewhere else that is in a better position to handle it. I'd be open to logging this, though I don't believe we have access to the log object at this point. We would need to either accept it in the constructor or cache it on the first request.

Reply:
After a conversation with @DonutEspresso, we settled on exposing this module's state through https://github.com/restify/node-restify/pull/1460/files#diff-01a30fd8ce166416dd691d3498c746c7R144

This allows for tooling outside of this plugin/repo to check and log/monitor/etc. the behavior of this plugin. When  fails, we can can check `cpu === NaN` in that module and log it out.